### PR TITLE
NAS-113296 / 22.02 / NAS-113296: Refactor Users to new ix table

### PR DESCRIPTION
**Pull request by denysbutenko**

### CODE REVIEW

File: src/app/pages/account/users/user-list/user-list.component.html [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            <mat-icon>
+              {{ expandedRow === user ? 'expand_less' : 'expand_more' }}
+            </mat-icon>
+          </button>
+        </td>
+      </ng-container>
+
+      <ng-container matColumnDef="expandedDetail">
+        <td mat-cell *matCellDef="let user; dataSource: dataSource" [attr.colspan]="displayedColumns.length">
+          <div class="ix-table-details" [@detailExpand]="user === expandedRow ? 'expanded' : 'collapsed'">
```
Comment (undsoft) [inline src/app/pages/account/users/user-list/user-list.component.html:57]: I guess we could consider extracting everything in this cell to a small component - will make testing easier.

File: src/app/pages/account/users/user-list/user-list.component.html [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            </section>
+
+            <mat-divider></mat-divider>
+
+            <div class="actions" fxLayout="row" fxLayoutGap="8px">
+              <button
+                mat-button
+                type="button"
+                (click)="doEdit(user)"
+                aria-label="Edit User"
```
Comment (undsoft) [inline src/app/pages/account/users/user-list/user-list.component.html:77]: aria-label can probably go

File: src/app/pages/account/users/user-list/user-list.component.html [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                  <span>{{ 'Delete' | translate }}</span>
+                </div>
+              </button>
+            </div>
+          </div>
+        </td>
+      </ng-container>
+
+      <tr mat-header-row *matHeaderRowDef="displayedColumns, sticky: true"></tr>
+      <tr mat-row *matRowDef="let user; columns: displayedColumns;"
```
Comment (undsoft) [inline src/app/pages/account/users/user-list/user-list.component.html:103]: Inconsistent formatting.

**Review by undsoft — COMMENTED** [review]
Got this weird behavior where rows don't open on the first try.
https://user-images.githubusercontent.com/1481270/144383932-6a9df9e8-b00c-485c-895d-9b86a7154b6f.mov

File: src/app/pages/common/ix-tables/components/ix-table-expandable-row/ix-table-expandable-row.component.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+@Component({
+  selector: 'ix-table-expandable-row',
+  templateUrl: './ix-table-expandable-row.component.html',
+  styleUrls: ['./ix-table-expandable-row.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  animations: [
+    trigger('rowExpand', [
+      state('collapsed', style({ height: '0px', minHeight: '0' })),
+      state('expanded', style({ height: '*' })),
+      transition('expanded <=> collapsed', animate('225ms cubic-bezier(0.4, 0.0, 0.2, 1)')),
```
Comment (dszidi) [inline src/app/pages/common/ix-tables/components/ix-table-expandable-row/ix-table-expandable-row.component.ts:18]: Animations on my system seem to feel rather sluggish. Maybe try a shorter animation time frame. 50ms yielded better results

Comment (denysbutenko) [reply src/app/pages/common/ix-tables/components/ix-table-expandable-row/ix-table-expandable-row.component.ts:18]: Did you click on the row itself or on expand button? I see a difference between them in the smoothness. Good catch, btw!

File: src/app/pages/common/ix-tables/components/ix-table-paginator/ix-table-paginator.component.ts [hunk 5]
```diff
@@ -41,7 +41,7 @@ export class IxTablePaginatorComponent<T> extends MatPaginator {
   }
 
   get total(): number {
-    return this.dataSource.data.length || 0;
+    return this.dataSource?.data.length || 0;
```
Comment (undsoft) [inline src/app/pages/common/ix-tables/components/ix-table-paginator/ix-table-paginator.component.ts:44]: If data is undefined, wouldn't this still give an error of: "Cannot read length of undefined"?

**Review by undsoft — CHANGES_REQUESTED** [review]

File: src/app/pages/account/users/user-list/user-list.component.html [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        <th mat-header-cell *matHeaderCellDef mat-sort-header>{{ 'Home' | translate }}</th>
+        <td mat-cell *matCellDef="let user; dataSource: dataSource">{{ user.home }}</td>
+      </ng-container>
+
+      <ng-container matColumnDef="full_name">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header>{{ 'Full Name' | translate }}</th>
+        <td mat-cell *matCellDef="let user; dataSource: dataSource">{{ user.full_name }}</td>
+      </ng-container>
+
+      <ng-container matColumnDef="actions">
```
[the remark this answers is not in the document]
Comment (denysbutenko) [reply src/app/pages/account/users/user-list/user-list.component.html:38]: I am not sure. What is the benefit of this?

For example:
```
<ix-table-actions-column>
<button ...></button>
</ix-table-actions-column>
```

It wouldn't reduce a lot of code but we will lose type checking when `*matCellDef` is inside its own component

[the remark this answers is not in the document]
Comment (undsoft) [reply src/app/pages/account/users/user-list/user-list.component.html:38]: I was thinking something along the lines of:
```
<ix-expand-toggle-column
  (toggle)="expandRow(user)"
  [expandedRow]="expandedRow"
></ix-expand-toggle-column>
```

The type safety inside will be reduced, but we won't be accessing any properties, only comparing references, can use generics or unknown for that.

[the remark this answers is not in the document]
Comment (denysbutenko) [reply src/app/pages/account/users/user-list/user-list.component.html]: done

File: src/app/pages/account/users/user-list/user-list.component.html [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+            <mat-icon>
+              {{ expandedRow === user ? 'expand_less' : 'expand_more' }}
+            </mat-icon>
+          </button>
+        </td>
+      </ng-container>
+
+      <ng-container matColumnDef="expandedDetail">
+        <td mat-cell *matCellDef="let user; dataSource: dataSource" [attr.colspan]="displayedColumns.length">
+          <ix-table-expandable-row [data]="getDetails(user)" [isExpanded]="expandedRow === user">
```
[the remark this answers is not in the document]
Comment (denysbutenko) [reply src/app/pages/account/users/user-list/user-list.component.html]: almost done, need to fix/write unit tests

File: src/app/pages/account/users/user-list-details/user-list-details.component.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+import { IxTableComponent } from 'app/pages/common/ix-tables/components/ix-table/ix-table.component';
+import { ModalService } from 'app/services';
+import { DialogService } from 'app/services/dialog.service';
+import { WebSocketService } from 'app/services/ws.service';
+
+@UntilDestroy()
+@Component({
+  selector: 'app-user-list-details',
+  templateUrl: './user-list-details.component.html',
+  styleUrls: ['./user-list-details.component.scss'],
```
Comment (undsoft) [inline src/app/pages/account/users/user-list-details/user-list-details.component.ts:25]: Empty file.

File: src/app/pages/account/users/user-list-details/user-list-details.component.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+export class UserListDetailsComponent implements OnInit, OnDestroy {
+  @Input() expandedRow: User;
+  @Input() dataSource: MatTableDataSource<User>;
+  @Input() colspan: number;
+  @Input() users: [User[]?] = [];
+  @Output() update = new EventEmitter<unknown>();
+  @ViewChild(MatRowDef, { static: false }) rowDef: MatRowDef<User>;
+  @ViewChild(MatColumnDef, { static: false }) columnDef: MatColumnDef;
+
+  groups: [Group[]?] = [];
```
Comment (undsoft) [inline src/app/pages/account/users/user-list-details/user-list-details.component.ts:37]: Looks like we can save on a request by only querying for a group when user is about to be deleted.

File: src/app/pages/common/ix-tables/components/ix-expand-toggle-column/ix-expand-toggle-column.component.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  Component,
+  Output, EventEmitter, Input, ViewChild, OnDestroy, OnInit, ChangeDetectorRef, ChangeDetectionStrategy,
+} from '@angular/core';
+import { MatColumnDef } from '@angular/material/table';
+import { IxTableComponent } from 'app/pages/common/ix-tables/components/ix-table/ix-table.component';
+
+@Component({
+  selector: 'ix-expand-toggle-column',
+  templateUrl: './ix-expand-toggle-column.component.html',
+  styleUrls: ['./ix-expand-toggle-column.component.scss'],
```
Comment (undsoft) [inline src/app/pages/common/ix-tables/components/ix-expand-toggle-column/ix-expand-toggle-column.component.ts:11]: Empty file.

File: src/app/pages/account/users/user-list-details/user-list-details.component.ts [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+  templateUrl: './user-list-details.component.html',
+  styleUrls: ['./user-list-details.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class UserListDetailsComponent implements OnInit, OnDestroy {
+  @Input() expandedRow: User;
+  @Input() dataSource: MatTableDataSource<User>;
+  @Input() colspan: number;
+  @Input() users: [User[]?] = [];
+  @Output() update = new EventEmitter<unknown>();
```
Comment (undsoft) [inline src/app/pages/account/users/user-list-details/user-list-details.component.ts:33]: Unused. If needs to be used, `EventEmitter<void>` may also work.

File: src/app/pages/account/users/user-list-details/user-list-details.component.html [hunk 12]
```diff
@@ -0,0 +1,37 @@
+<ng-container matColumnDef="expandedDetail">
+  <td mat-cell *matCellDef="let user; dataSource: dataSource" [attr.colspan]="colspan">
+    <ix-table-expandable-row [data]="getDetails(user)" [isExpanded]="expandedRow === user">
```
Comment (undsoft) [inline src/app/pages/account/users/user-list-details/user-list-details.component.html:3]: Not sure about this, but initially I thought we'd only move what's inside of cell definition (rows 3 - 28) into `user-list-details` and keep rest (table related stuff) in parent component. This seems to fit single responsibility principle better, where parent component deals with tables and details component deals with a specific row.
But I am not super sure on this one, so we can keep current approach as well.

**Review by undsoft — APPROVED** [review]
Nice work!

**Pull request merged**